Extract ItemReconciler service

lucaswerkmeister opened the pull request:
This service wraps a `ReconciliationService` and `SimplePutStrategy` and extracts the reconciliation URL from an input item before passing it to the `ReconciliationService` and `SimplePutStrategy`. Its return value is an Item and a revision ID, but the meaning is quite different, so we introduce a new result class, `ReconciledItem`, rather than reusing `ReconciliationItem`.

Bug: [T283567](https://phabricator.wikimedia.org/T283567)

---

Includes one preparation commit to move some classes.

The names “ItemReconciler” and “ReconciledItem” aren’t that great… any other ideas?

lucaswerkmeister committed:
Move Reconciliation* classes out of MediaWiki namespace

ReconciliationService and its result ReconciliationItem move into a new
Reconciliation namespace; ReconciliationException moves into the
top-level extension namespace, since it’s interpreted as the base class
for all exceptions of the extension, and is in fact mostly thrown by
code in other namespaces.

lucaswerkmeister committed:
Extract ItemReconciler service

This service wraps a ReconciliationService and SimplePutStrategy and
extracts the reconciliation URL from an input item before passing it to
the ReconciliationService and SimplePutStrategy. Its return value is an
Item and a revision ID, but the meaning is quite different, so we
introduce a new result class, ReconciledItem, rather than reusing
ReconciliationServiceItem.

Bug: T283567

toban commented on includes/MediaWiki/Api/EditEndpoint.php:
```diff
@@ -153,14 +111,14 @@ public function run() {
 				$toSave,
 				'Reconciliation Edit',
-				$reconciliationItem->getRevision() ? null : EDIT_NEW,
+				$reconciledItem->isNew() ? EDIT_NEW : 0,
```
Shouldn't this be `EDIT_UPDATE`?

Seems like the 0 or null becomes a `EDIT_AUTOSUMMARY` by default? Not sure if it even matters.

lucaswerkmeister replied:
Hm, `EDIT_UPDATE` might make sense. But where do you see `EDIT_AUTOSUMMARY` coming in?

toban replied:
I thought it was set here https://github.com/wikimedia/Wikibase/blob/REL1_35/repo/includes/EditEntity/MediawikiEditEntity.php#L739 if left out

lucaswerkmeister replied:
Okay, but it doesn’t look like we can influence this, so I’d leave it out of this call.

toban commented on includes/Reconciliation/ReconciliationItem.php:
```diff
@@ -0,0 +1,48 @@
+ *
+ * @license GPL-2.0-or-later
+ */
+class ReconciliationItem {
```
How do you feel about naming this `ReconciliationServiceItem` ?

Even though we have the comment not to confuse the two i think making the difference a bit more obvious will make this less likely to be confused?

lucaswerkmeister replied:
Sounds okay to me. Or maybe even `ReconciliationServiceResult`, without “item”?

lucaswerkmeister replied:
Okay, I’ll insert another commit to rename it to `ReconciliationServiceResult`.

lucaswerkmeister replied:
Hm, on the other hand the term “item” is used a lot for these results (e.g. `EditRequest::otherItems()`), and changing those to “result” would probably make it less clear overall.

Let’s go with `ReconciliationServiceItem`.

toban reviewed (approved):
Looks good to me, just a few thoughts that popped up.

lucaswerkmeister committed:
Rename ReconciliationItem to ReconciliationServiceItem

This is intended to distinguish it better from the upcoming
ReconciledItem class. Related variables are also renamed accordingly,
sometimes to “reconciliation service items”, sometimes to “items”; in
ReconciliationServiceTest, to avoid the term “reconciled items”, rename
a parameter to $previousResults instead.
